lib/auth/password.go

```diff
@@ -178,23 +257,110 @@ func (s *AuthServer) getOTPType(user string) (string, error) {
 	return teleport.HOTP, nil
 }
 
-// GetOTPData returns the OTP Key, Key URL, and the QR code.
-func (s *AuthServer) GetOTPData(user string) (string, []byte, error) {
-	// get otp key from backend
-	otpSecret, err := s.GetTOTP(user)
+func (s *AuthServer) changePasswordWithToken(req ChangePasswordWithTokenRequest) (services.User, error) {
+	clusterConfig, err := s.GetClusterConfig()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if clusterConfig.GetLocalAuth() == false {
+		return nil, trace.AccessDenied(noLocalAuth)
+	}
+
+	err = services.VerifyPassword(req.Password)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	userToken, err := s.GetUserToken(req.TokenID)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if userToken.Expiry().Before(s.clock.Now().UTC()) {
+		return nil, trace.BadParameter("expired token")
+	}
+
+	username := userToken.GetUser()
+	err = s.changeUserSecondFactor(req, userToken)
 	if err != nil {
-		return "", nil, trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
 
-	// create otp url
-	params := map[string][]byte{"secret": []byte(otpSecret)}
-	otpURL := utils.GenerateOTPURL("totp", user, params)
+	err = s.deleteUserTokens(username)
```

Comment:
Shouldn't the users second factor and password be updated, then the token deleted?

Reply:
Since these are not part of one atomic operation, it's better to delete a token first before issuing DB writes as we do not want to end up with a valid token + partially updated user.